src/olympia/discovery/tests/test_admin.py

```diff
@@ -546,6 +552,156 @@ def test_query_count(self):
+        response = self.client.get(self.list_url, follow=True)
+        assert response.status_code == 200
+        assert 'föo.jpg' in response.content.decode('utf-8')
+
+    def test_can_edit_with_discovery_edit_permission(self):
+        item = PrimaryHeroImage.objects.create(custom_image='föo.jpg')
+        self.detail_url = reverse(
+            'admin:discovery_primaryheroimageupload_change', args=(item.pk,)
+        )
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(self.detail_url, follow=True)
+        assert response.status_code == 200
+        content = response.content.decode('utf-8')
+        assert u'föo.jpg' in content
+
+        updated_photo = get_uploaded_file('transparent.png')
+        response = self.client.post(
+            self.detail_url,
+            dict(custom_image=updated_photo),
+            follow=True)
+        assert response.status_code == 200
+        item.reload()
+        assert PrimaryHeroImage.objects.count() == 1
+        assert item.custom_image == 'hero-featured-image/transparent.png'
+
+    def test_can_delete_with_discovery_edit_permission(self):
+        item = PrimaryHeroImage.objects.create()
+        delete_url = reverse(
+            'admin:discovery_primaryheroimageupload_delete', args=(item.pk,)
+        )
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        # Can access delete confirmation page.
+        response = self.client.get(delete_url, follow=True)
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.filter(pk=item.pk).exists()
+
+        # And can actually delete.
+        response = self.client.post(
+            delete_url, data={'post': 'yes'}, follow=True)
+        assert response.status_code == 200
+        assert not PrimaryHeroImage.objects.filter(pk=item.pk).exists()
+
+    def test_can_add_with_discovery_edit_permission(self):
+        add_url = reverse('admin:discovery_primaryheroimageupload_add')
+        user = user_factory()
+        self.grant_permission(user, 'Admin:Tools')
+        self.grant_permission(user, 'Discovery:Edit')
+        self.client.login(email=user.email)
+        response = self.client.get(add_url, follow=True)
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.count() == 0
+        photo = get_uploaded_file('transparent.png')
+        response = self.client.post(
+            add_url,
+            dict(custom_image=photo),
+            follow=True)
+        assert response.status_code == 200
+        assert PrimaryHeroImage.objects.count() == 1
+        item = PrimaryHeroImage.objects.get()
+        assert item.custom_image == 'hero-featured-image/transparent.png'
```

Comment:
Can you assert that the file now exists in the uploads folder?  `assert os.path.exists(os.path.join(settings.MEDIA_ROOT, 'hero-featured-image', 'transparent.png'))` I guess.